praw/models/reddit/rules.py

```diff
@@ -0,0 +1,296 @@
+                violation_reason="Spam post')
+
+        """
+        data = {"r": str(self.subreddit), "old_short_name": self.short_name}
+        for name, value in {
+            "description": description,
+            "kind": kind,
+            "short_name": short_name,
+            "violation_reason": violation_reason,
+        }.items():
+            data[name] = getattr(self, name) if value is None else value
+        updated_rule = self._reddit.request(
+            "POST", API_PATH["update_subreddit_rule"], data=data
+        )
+        self._reddit._objector.objectify(updated_rule)
+        # We want any errors to be passed, but we don't want to use the
+        # objected value because it returns a LiveThread.
+        rule_data = loads(updated_rule["json"]["data"]["rules"])[0]
+        return Rule(self._reddit, self.subreddit, _data=rule_data)
+
+
+class SubredditRules:
+    """Provide a set of functions to a Subreddit's rules."""
+
+    def __getitem__(self, short_name: Union[str, int]) -> Rule:
+        """Return the Rule for the subreddit with short_name ``short_name``.
+
+        :param short_name: The short_name of the rule, or the rule number.
+
+        .. note:: Rules fetched using a specific rule name are lazy loaded, so
+            you might have to access an attribute to get all of the expected
+            attributes.
+
+        This method is to be used to fetch a specific rule, like so:
+
+        .. code-block:: python
+
+            rule_name = 'No spam'
+            rule = reddit.subreddit('NAME').rules[rule_name]
+            print(rule)
+
+        You can also fetch a numbered rule of a subreddit.
+
+        Rule numbers start at 0, so the first rule is at index 0, and the
+        second rule is at index 1, and so on.
+
+        If a rule of a specific number does not exist, an
+        :py:class:`IndexError` will be thrown.
+
+        .. note:: You can use negative indexes, such as -1 to get the last rule.
+
+        For example, to fetch the second rule of ``AskReddit``:
+
+        .. code-block:: python
+
+            rule = reddit.subreddit("AskReddit").rules[1]
+
+        """
+        if isinstance(short_name, int):
+            return list(self)[short_name]
+        return Rule(self._reddit, self.subreddit, short_name)
+
+    def __init__(self, subreddit: Subreddit):
+        """Create a SubredditRules instance.
+
+        :param subreddit: The subreddit whose rules to work with.
+
+        """
+        self.subreddit = subreddit
+        self._reddit = subreddit._reddit
+
+    def __iter__(self) -> Generator[Rule, None, None]:
+        """Return a list of rules for the subreddit.
+
+        :returns: A generator containing all of the rules of a subreddit.
+
+        This method is used to discover all rules for a subreddit:
+
+        .. code-block:: python
+
+           for rule in reddit.subreddit('NAME').rules:
+               print(rule)
+
+        """
+        for rule in self._make_rule_list(
```

Comment:
Let's make this be a lazy-loaded and cached property.

Here's how I suggest this happens:

- Rename the `_make_rule_list` method to just `_rule_list` and use the `@cachedproperty` decorator ([example here](https://github.com/praw-dev/praw/blob/1a78ea82c9cac3c413b4a71da94991ef29636584/praw/models/listing/mixins/subreddit.py#L53)). Then, the rules won't be loaded until `self._rule_list` is accessed, and once that property is accessed once, its value will be saved so that we don't have to make another network request.
- Update `_rule_list` to not take any parameters and perform the `get` that is currently performed in `__iter__`. (Note: I see that you pass something else in as this argument, but I suspect this will be avoided once we update the objector to support `Rule`s.)
- Change `__iter__` to just iterate through `self._rule_list` and yield its items.

This also resolves my comment above in `__getitem__` because you can change that to `return self._rule_list[short_name]`.

Reply:
This seems like a promising idea, but that the cache needs to get rebuilt whenever a method is called on any rule. More specifically, if I re-arrange rules, or add a rule, the rule list won't get updated. Per the spirit of PRAW, I don't think we should cache anything that makes network requests.

Reply:
> the cache needs to get rebuilt whenever a method is called on any rule

Could be useful, but this differs from existing behavior, at least in some places. Many PRAW classes let you update them and don't refresh without that being specified.

> I don't think we should cache anything that makes network requests

I'm not sure what you mean by this. We cache comment bodies even though they can be edited.

Reply:
What I'm saying is that if I delete a rule, that should get reflected. Are you saying that in order to get a new rule list, the subreddit object would need to be re-created?

Reply:
> Are you saying that in order to get a new rule list, the subreddit object would need to be re-created?

This is how it tends to operate in other classes. But I would be down for invalidating the cached response whenever we update or delete a rule.

Reply:
I think this is going to require a flag in the parent Subreddit class.